File: cpp/comparaison_somme.cpp
#include <iostream>
#include <chrono>

static long iterSum(long until)
{
    long total = 0;
    while (until)
        total += until--;
    return total;
}

static long recursSum(long until)
{
    if (until == 1)
        return 1;
    else
        return 1 + recursSum(until - 1);
}

int main(void) {
    long result = 0;
    auto iterStart = std::chrono::steady_clock::now();
    result = iterSum(150);
    auto iterEnd = std::chrono::steady_clock::now(); 
    auto iterNs = iterEnd - iterStart;
    
    auto recursStart = std::chrono::steady_clock::now();
    result = recursSum(150);
    auto recursEnd = std::chrono::steady_clock::now();
    auto recursNs = recursEnd - recursStart;
    
    std::cout << "la fonction itérative à pris " << iterNs.count()
              << " nanosecondes, la fonction récursive a pris " << recursNs.count()
              << " nanosecondes.\n";

    return 0;
}

File: js/comparaison_somme.js
function iterSum(until) {
    let total = 0;
    while (until > 0) {
        total += until--;
    }
    return total;
}

function recursSum(until) {
    if (until === 1) {
        return 1;
    } else {
        return until + recursSum(until - 1);
    }
}

function main() {
    let result = 0;

    let iterStart = performance.now();
    result = iterSum(150);
    let iterEnd = performance.now();
    let iterNs = iterEnd - iterStart;

    let recursStart = performance.now();
    result = recursSum(150);
    let recursEnd = performance.now();
    let recursNs = recursEnd - recursStart;

    console.log("la fonctions itérative a pris " + iterNs + " millisecondes, la fonction récursive a pris " + recursNs + " millisecondes.");
}

main();
